pkg/xds/envoy/listeners/fault_injection_configurer.go

```diff
@@ -0,0 +1,129 @@
+package listeners
+
+import (
+	envoy_listener "github.com/envoyproxy/go-control-plane/envoy/api/v2/listener"
+	envoy_api_v2_route "github.com/envoyproxy/go-control-plane/envoy/api/v2/route"
+	envoy_filter_fault "github.com/envoyproxy/go-control-plane/envoy/config/filter/fault/v2"
+	envoy_http_fault "github.com/envoyproxy/go-control-plane/envoy/config/filter/http/fault/v2"
+	envoy_hcm "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	envoy_type_matcher "github.com/envoyproxy/go-control-plane/envoy/type/matcher"
+	envoy_wellknown "github.com/envoyproxy/go-control-plane/pkg/wellknown"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/wrappers"
+
+	mesh_proto "github.com/Kong/kuma/api/mesh/v1alpha1"
+	"github.com/Kong/kuma/pkg/xds/envoy/routes"
+	"github.com/Kong/kuma/pkg/xds/envoy/tags"
+)
+
+func FaultInjection(faultInjection *mesh_proto.FaultInjection) FilterChainBuilderOpt {
+	return FilterChainBuilderOptFunc(func(config *FilterChainBuilderConfig) {
+		config.Add(&FaultInjectionConfigurer{
+			faultInjection: faultInjection,
+		})
+	})
+}
+
+type FaultInjectionConfigurer struct {
+	faultInjection *mesh_proto.FaultInjection
+}
+
+func (f *FaultInjectionConfigurer) Configure(filterChain *envoy_listener.FilterChain) error {
+	if f.faultInjection == nil {
+		return nil
+	}
+
+	config := &envoy_http_fault.HTTPFault{
+		Delay: convertDelay(f.faultInjection.Conf.GetDelay()),
+		Abort: convertAbort(f.faultInjection.Conf.GetAbort()),
+		Headers: []*envoy_api_v2_route.HeaderMatcher{
+			createHeaders(f.faultInjection.SourceTags()),
+		},
```

Comment:
Please remind me. Why can't we generate separate `envoy_api_v2_route.HeaderMatcher` for every selector.Match from FaultInjection? This would simplify the logic of constructing gigantic regex with ORs to construct regex for every selector.Match

Reply:
I discovered that `HeaderMatcher` (at least with the same header's name) works as AND. We also have an option to generate separate fault filters, but that's probably bulkier solution

Reply:
Thanks, you are right
"A match will happen if all the headers in the config are present in the request with the same values (or based on presence if the value field is not in the config)."
https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/http/fault/v2/fault.proto

I'd lean towards picking a solution with generating many fault filters just to avoid crazy long regexes (btw. is 500 the limit of chars in regex?) and also for easier migration to `x-kuma-tag-*` in the future, but we can stick to or regex solution as well for now.

Reply:
This is not exactly chars in regex:

_This field controls the RE2 “program size” which is a rough estimate of how complex a compiled regex is to evaluate. A regex that has a program size greater than the configured value will fail to compile. In this case, the configured max program size can be increased or the regex can be simplified. If not specified, the default is 100._

But there is a linear dependency between the size in chars and "program size". 
In both cases (many filters or complex regex) we will be limited by that "program size" and limited with number of tags/selectors in the policy. 

I'd stay with long regex for this PR and later we can compare which approach is faster - e.g. 10 filter faults with program size 500 or one filter fault with program size 5000

Reply:
Oh and also we probably can calculate "program size" on the fly